### PR TITLE
Add Git Reset, Rebase Workflow article

### PR DESCRIPTION
**Pull request by kapowaz**
Whaddya know, it’s an article where I talk about the health benefits of using `git rebase`!

img

### CODE REVIEW

File: _posts/2017-09-07-reset-rebase-workflow.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+---
+
+Small, atomic commits are the gold-standard for a nice, clean git commit history. If you embrace this ideology in your working practice you may find yourself committing frequently, but whilst you're in the early stages of developing a feature it's quite possible you'll end up making decisions that you reverse-course on, or tweak as the concept solidifies.
+
+Potentially this can mean you end up with a commit history containing many small ‘work in progress' commits which you don't necessarily want to merge into master in their current form. Sometimes you'll make a change, then decide to solve a problem an entirely different way, undoing that change later before your feature branch is even merged. You may also end up writing code that — whilst it _supports_ your work — has no direct requirement that it be _part of_ the feature you're building.
```
Comment (Crunch09) [inline _posts/2017-09-07-reset-rebase-workflow.md]: ‘work in progress' => 'work in progress'

File: _posts/2017-09-07-reset-rebase-workflow.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    Implement new widget
+
+commit 707548c16a44f629584fc4bae93b8014888c19a8
+Author: Ben Darlow <ob.fusc@t.ed>
+Date:   Thu Sep 7 14:29:04 2017 +0100
+
+    Initial commit
+```
+
+Once you're happy with the state of this new feature branch, you can safely delete the old `new-widget-search` branch, and rename the newer branch by removing `-rebased` from the end. If you already pushed the branch, you'll also want to ensure you set the remote URL for this newer branch to the existing one, then _force push_ to overwrite the old branch. Of course, all the usual caveats about using force push apply: if anybody else is working on this branch too, make sure you talk to them before you force push and wipe out their changes!
```
Comment (Crunch09) [inline _posts/2017-09-07-reset-rebase-workflow.md]: > rename the newer branch by removing `-rebased` from the end

Don't quite understand what you mean here?

**Comment by evelyne24** [comment]
@kapowaz I am a huge fan of rebase, I do the exact same things you wrote in your article, including using Git Tower 2 but also command line with [oh-my-zsh Git](https://github.com/robbyrussell/oh-my-zsh/wiki/Cheatsheet) plugin shortcuts. One thing I do to achieve clean cut commits is to go back and modify existing commits.
For example, let's say I have clean commits A, B, C and now I realised I forgot a change that would go well with commit B. I can create a new commit D, to say something like "Forgot to add a test for the change I made in B" but this will make it hard for people to follow if they use the "commit-by-commit" approach (which people do a LOT, in the Android team)

What I tend to do is (with the oh-my-zsh shortcuts):

```bash
glgg # or git log --graph --max-count = 10
```
I see the tree of commits:

```
* commit f442511407a73d51ab9c974c38d21b77d1ea6d8d
| Author: Evelina Vrabie <evelina.vrabie@deliveroo.co.uk>
| Date:   Fri Sep 8 11:15:55 2017 +0100
|
|     Commit C
|
* commit 6724df5805295198006d00079aebfb516e656df7
| Author: Evelina Vrabie <evelina.vrabie@deliveroo.co.uk>
| Date:   Fri Sep 8 10:45:31 2017 +0100
|
|     Commit B
|
* commit cafbc74551945e0ece617fd4720466a80f0ea88e
| Author: Evelina Vrabie <evelina.vrabie@deliveroo.co.uk>
| Date:   Thu Sep 7 17:16:16 2017 +0100
|
|     Commit A
|
*   commit 0f0e7a97e554085722daad685737b86dd74c15ed
|\  Merge: f7ebe2cdb cbbd464a9
| | Author: Ana Capatina <anikiki@users.noreply.github.com>
| | Date:   Thu Sep 7 15:04:43 2017 +0100
| |
| |     Merge pull request #1859 from deliveroo/CCC-173-unavailable-items-can-be-added-to-basket
| |
| |     [CCC-173] [timebox: 1 day] Unavailable items can be added to basket
| |
:
```

I rebase on the commit before the one I want to edit:

```bash
grbi cafbc74551945e0ece617fd4720466a80f0ea88e # or git rebase -i cafbc74551945e0ece617fd4720466a80f0ea88e
```

Now I have options for my commit:

```bash
pick 6724df580 Commit B
pick f44251140 Commit C

 # Rebase 0f0e7a97e..f44251140 onto 0f0e7a97e (3 commands)
 #
 # Commands:
 # p, pick = use commit
 # r, reword = use commit, but edit the commit message
 # e, edit = use commit, but stop for amending
 # s, squash = use commit, but meld into previous commit
 # f, fixup = like "squash", but discard this commit's log message
 # x, exec = run command (the rest of the line) using shell
 # d, drop = remove commit
 #
 # These lines can be re-ordered; they are executed from top to bottom.
 #
 # If you remove a line here THAT COMMIT WILL BE LOST.
 #
 # However, if you remove everything, the rebase will be aborted.
 #
 # Note that empty commits are commented out
 ```

I chose to edit the commit I want, which is B:
```bash
e 6724df580 Commit B
pick f44251140 Commit C
```

It now tells me:
```bash
Stopped at 6724df580... CCNR-263 Start add address flow instead of showing the map
You can amend the commit now, with

	git commit --amend

Once you are satisfied with your changes, run

	git rebase --continue
```

I make my changes, then add the changes:

```bash
ga <files changed> # or git add <files changed>
gc --amend # or git commit --ammend 
```

I verify the commit description and save then continue the rebase.

```bash
grbc # or git rebase --continue
```

I do this dozens of times a day and having those aliases helped me speed up things a lot.

File: _posts/2017-09-07-reset-rebase-workflow.md (hunk 2 above)
Comment (kapowaz) [reply _posts/2017-09-07-reset-rebase-workflow.md]: Yeah, I struggled to phrase this in a way that felt right. Earlier in the instructions I gave the command to create a branch called `new-widget-search-rebased` and so this is about renaming the branch without the `-rebased` suffix. I’ll have a think and reword it 👍

Comment (kapowaz) [reply _posts/2017-09-07-reset-rebase-workflow.md]: I’m thinking maybe:

> Once you're happy with the state of this new feature branch, you can safely delete the old branch, and rename the newer branch to `new-widget-search` in its place.

What do you think?

**Comment by kapowaz** [comment]
@evelyne24 that’s amazing. Stuff like this definitely needs to be talked about more — I feel like a lot of people interface with git as little as possible (this was me a few years ago!), but once you get the hang of these more advanced commands you can do so much with it!

**Comment by evelyne24** [comment]
Well...I kind of assumed people knew these things already and nobody wants me to blab about them but see...having a blog helps to surface that people are actually interested in these things :D

File: _posts/2017-09-07-reset-rebase-workflow.md (hunk 2 above)
Comment (kapowaz) [reply _posts/2017-09-07-reset-rebase-workflow.md]: Actually I think I’ll just add some more example commands to make it more explicit, I think that should clear it up.

File: _posts/2017-09-07-reset-rebase-workflow.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+## If we are to be prepared for it, we must first shed our fear of it
+
+<figure>
+![I stand here, before you now, truthfully unafraid](/images/posts/reset-rebase-workflow/morpheus.jpg)
+</figure>
+
+To paraphrase [Angus Croll][angus-croll-javascript], _you can't master a tool until you know it inside out – and fear and evasion are the enemies of knowledge_, so I encourage you _not_ to be [scared of git rebase][dont-be-scared-of-git-rebase]! There are numerous benefits to the workflow I'm about to describe, which you too can enjoy by conquering your fear of `git rebase`:
+
+- [They're easier to review (and _get_ reviewed!)][pr-roulette]
```
Comment (evelyne24) [inline _posts/2017-09-07-reset-rebase-workflow.md]: Is the link correct?

File: _posts/2017-09-07-reset-rebase-workflow.md (hunk 2 above)
Comment (Crunch09) [reply _posts/2017-09-07-reset-rebase-workflow.md]: 👍

File: _posts/2017-09-07-reset-rebase-workflow.md (hunk 3 above)
Comment (kapowaz) [reply _posts/2017-09-07-reset-rebase-workflow.md]: Yeah, I’m using a Markdown [reference link](https://daringfireball.net/projects/markdown/syntax#link) — it points to [this line](https://github.com/deliveroo/deliveroo.engineering/pull/157/files/d82723d6664183573544619dcd1be75a0f95bb9c#diff-5f5544e3199b46e87d0913484bb81e94R200)

Comment (evelyne24) [reply _posts/2017-09-07-reset-rebase-workflow.md]: Ah, wow, that's cool!

**Comment by kapowaz** [comment]
I’ve pushed some minor changes:

* Updated the date of publication to 2nd October 2017 (the existing publication date of the 7th September was kind of a while ago…)
* Used a different _Rick & Morty_ image that seemed more apt 😁

**Pull request merged**